### PR TITLE
DBZ-2897 Extended RecordCommitter interface to handle updated source offsets.

### PR DESCRIPTION
**Pull request by twthorn**
https://issues.redhat.com/browse/DBZ-2897

Extended the DebeziumEngine interface to include a new method for marking a record processed with updated source offsets. Implemented the new interface method in both ConvertingEngineBuilder and EmbeddedEngine. EmbeddedEngine creates a new SourceRecord with the updated sourceOffsets and passes this to the original markProcessed method. Added a unit test in EmbeddedEngineTest that verifies that the file offset storage contains the updated source offset’s partition. Added a revapi ignore case to prevent failure from a method added to an interface without a default method. Updated docs for Debezium Engine that describe the new functionality.

### CODE REVIEW

**Comment by jpechane** [comment]
@gunnarmorling LGTM. The only question I have is whether the sources should be encapsulated into an another datatype effectivelly decoupling from the map but I am not really sure how this could be done.

File: debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java [hunk 1]
```diff
@@ -328,6 +331,80 @@ public void shouldRunDebeziumEngine() throws Exception {
...
+        try (BufferedReader br = new BufferedReader(new FileReader(OFFSET_STORE_PATH.toString()))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                logger.info(line);
+                if (line.contains(CUSTOM_SOURCE_OFFSET_PARTITION)) {
+                    containsCustomPartition = true;
+                }
+            }
+        }
+        assert containsCustomPartition;
```
Comment (gunnarmorling) [inline debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java:402]: Should use `assertThat()`.

**Review by gunnarmorling — COMMENTED** [review]
One comment inline.

Re the offset representation as map, yeah, good point. We could have a simple contract like this instead:

```
interface Offsets {
    Offsets set(String key, Object value);
}
```

And pass in that instead of the map.

**Comment by twthorn** [comment]
Updated to use assertThat rather than java assert keyword.

Should I also add the `Offsets` contract in this pull request? If so, could you provide some more info on where you’d want it defined, and where it would be implemented (e.g., whether by the client or by `EmbeddedEngine`)? Would this follow a similar pattern to `RecordChangeEvent` and `EmbeddedEngineChangeEvent`? Any additional info would be helpful.

File: debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java (hunk 1 above)
Comment (twthorn) [reply debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java]: Good point, thank you, updated.

**Comment by gunnarmorling** [comment]
> Should I also add the Offsets contract in this pull request?

Yes, please.

> some more info on where you’d want it defined

I'd suggest an inner interface in `DebeziumEngine`, like `RecordCommitter` itself. The implementation would live next to where that one is implemented, too.

**Comment by twthorn** [comment]
Added `Offsets` contract in `DebeziumEngine` and implemented it as `SourceRecordOffsets` in `EmbeddedEngine`. Added additional method `buildOffsets` in `RecordCommitter` for retrieving an `Offsets` object, implemented in `ConvertingEngineBuilder` and `EmbeddedEngine`. Modified `markProcessed` method to use `Offsets` contract. Updated tests and documentation.

File: debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java [hunk 2]
```diff
@@ -328,6 +330,81 @@ public void shouldRunDebeziumEngine() throws Exception {
...
+            Thread.sleep(10);
+        }
+        allLatch.await(5000, TimeUnit.MILLISECONDS);
+        assertThat(allLatch.getCount()).isEqualTo(0);
+
+        boolean containsCustomPartition = false;
+        try (BufferedReader br = new BufferedReader(new FileReader(OFFSET_STORE_PATH.toString()))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                logger.info(line);
```
Comment (gunnarmorling) [inline debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java:396]: So the line that's read back here is some binary representation, not plain text string. We should be able to get back the proper offset by reading this as a byte array and feeding it through `JsonConverter#toConnectData()` (ie. inverting the offset serialization step). While the assertion below technically passes, it feels a bit weird, and you could also do a proper lookup in the offset map then.

File: debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java [hunk 3]
```diff
@@ -121,6 +121,36 @@ default void taskStopped() {
...
+         * @return the object implementing the {@link Offsets} contract
+         */
+        Offsets buildOffsets();
+    }
+
+    /**
+     * Contract that should be passed to {@link RecordCommitter#markProcessed(Object, Offsets)} for marking a record
+     * as processed with updated offsets.
+     */
+    public static interface Offsets {
```
Comment (gunnarmorling) [inline debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java:145]:
```suggestion
    public interface Offsets {
```

It doesn't make a difference here.

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot, @twthorn. Looks good to me implementation-wise now. One small nit re the test assertion only; other than that, it's good to go.

File: debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java (hunk 2 above)
Comment (twthorn) [reply debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java]: Good point, updated.

File: debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java (hunk 3 above)
Comment (twthorn) [reply debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java]: Updated.

**Comment by twthorn** [comment]
Made the offsets interface non-static. Refactored the unit test for updating offsets to instead read the offset file as a map of byte array to byte array, parse the byte array as a json string, deserialize the json string, and make assertions on the key/values in the deserialized JSON object. Also minor refactors for some test constants to make assertions clearer.

**Review by gunnarmorling — APPROVED** [review]

**Comment by gunnarmorling** [comment]
LGTM, thx! That test failure is unrelated and already tracked. Merging.

**Pull request merged**